src/lib/timing.c

```diff
@@ -0,0 +1,207 @@
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+   | Authors: Derick Rethans <derick@xdebug.org>                          |
+   |          Michael Voříšek <mvorisek@mvorisek.cz>                      |
+   +----------------------------------------------------------------------+
+ */
+
+#include "php_xdebug.h"
+#if PHP_WIN32
+	#include "win32/time.h"
+	#include <versionhelpers.h>
+#else
+	#include <sys/time.h>
+#endif
+#if __APPLE__
+	#include <mach/mach_time.h>
+#endif
+
+#include "timing.h"
+
+#define NANOTIME_MIN_STEP 10
+
+#if PHP_WIN32
+	#define WIN_NANOS_IN_TICK 100
+	#define WIN_TICKS_SINCE_1601_JAN_1 116444736000000000ULL
+#endif
+
+ZEND_EXTERN_MODULE_GLOBALS(xdebug)
+
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_init *nanotime_init)
+{
+	#if PHP_WIN32
+		FILETIME filetime;
+	#endif
+	#if _SC_MONOTONIC_CLOCK
+		struct timespec ts;
+	#endif
+	#if HAVE_GETTIMEOFDAY
+		struct timeval tp;
+	#endif
+
+	// Windows
+	#if PHP_WIN32
+		if (nanotime_init->win_precise_time_func != NULL) {
+			nanotime_init->win_precise_time_func(&filetime);
+			return ((((uint64_t)filetime.dwHighDateTime << 32) + (uint64_t)filetime.dwLowDateTime) - WIN_TICKS_SINCE_1601_JAN_1)
+				* WIN_NANOS_IN_TICK;
+		}
+	#endif
+
+	// Linux/Unix
+	#if _SC_MONOTONIC_CLOCK
+		if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
+			return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+		}
+
+		return 0;
+	#endif
+
+	// fallback if better platform specific time is not available
+	#if HAVE_GETTIMEOFDAY
+		if (gettimeofday(&tp, NULL) == 0) {
+			return (uint64_t)tp.tv_sec * NANOS_IN_SEC + (uint64_t)tp.tv_usec * NANOS_IN_MICROSEC;
+		}
+	#endif
+
+	return 0;
+}
+
+#if PHP_WIN32
+	static uint64_t xdebug_counter_and_freq_to_nanotime(uint64_t counter, uint64_t freq)
+	{
+		uint32_t mul = 1, freq32;
+		uint64_t q, r;
+
+		while (freq >= (1ULL << 32)) {
+			freq /= 2;
+			mul *= 2;
+		}
+		freq32 = (uint32_t)freq;
+
+		q = counter / freq32;
+		r = counter % freq32;
+		return (q * NANOS_IN_SEC + (r * NANOS_IN_SEC) / freq32) * mul;
+	}
+#endif
+
+static uint64_t xdebug_get_nanotime_rel(xdebug_nanotime_init *nanotime_init)
+{
+	// Windows Windows 7 and lower
+	#if PHP_WIN32
+		LARGE_INTEGER tcounter;
+
+		if (nanotime_init->win_precise_time_func == NULL) {
+			QueryPerformanceCounter(&tcounter);
+			return xdebug_counter_and_freq_to_nanotime((uint64_t)tcounter.QuadPart, nanotime_init->win_freq);
+		}
+	#endif
+
+	// Mac
+	// should be fast but can be relative
+	#if __APPLE__
+		return clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
+	#endif
+
+	return 0;
+}
+
+uint64_t xdebug_get_nanotime(void)
+{
+	uint64_t nanotime;
+	xdebug_nanotime_init *nanotime_init;
+
+	nanotime_init = &XG_BASE(nanotime_init);
+
+	nanotime = xdebug_get_nanotime_rel(nanotime_init);
+	if (nanotime > 0) {
```

Comment:
Instead of checking for the return `0`, can we not put this in a pre-process check `#if PHP_WIN32|__APPLE__` ?

Reply:
Windows is special, because on Windows 7 and below, we need relative time source because the absolute precision one is not available - and this is runtime based. How to fix this?